pkg/controller/aws/identity/iamrole.go

```diff
@@ -0,0 +1,182 @@
+		Named(name).
+		For(&v1alpha1.IAMRole{}).
+		Complete(r)
+}
+
+type roleConnector struct {
+	client      client.Client
+	newClientFn func(*aws.Config) (iam.RoleClient, error)
+}
+
+func (conn *roleConnector) Connect(ctx context.Context, mgd resource.Managed) (resource.ExternalClient, error) {
+	cr, ok := mgd.(*v1alpha1.IAMRole)
+	if !ok {
+		return nil, errors.New(errObjectNotIAMRole)
+	}
+
+	newClientFn := iam.NewRoleClient
+	if conn.newClientFn != nil {
+		newClientFn = conn.newClientFn
+	}
+
+	awsconfig, err := utils.RetrieveAwsConfigFromProvider(ctx, conn.client, cr.Spec.ProviderReference)
+	if err != nil {
+		return nil, err
+	}
+
+	c, err := newClientFn(awsconfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot create a new IAMRole client")
+	}
+	return &roleExternal{c}, nil
+}
+
+type roleExternal struct {
+	client iam.RoleClient
+}
+
+func (e *roleExternal) Observe(ctx context.Context, mgd resource.Managed) (resource.ExternalObservation, error) {
+	cr, ok := mgd.(*v1alpha1.IAMRole)
+	if !ok {
+		return resource.ExternalObservation{}, errors.New(errObjectNotIAMRole)
+	}
+
+	req := e.client.GetRoleRequest(&awsiam.GetRoleInput{
+		RoleName: aws.String(cr.Spec.RoleName),
+	})
+	req.SetContext(ctx)
+
+	observed, err := req.Send()
+
+	if iam.IsErrorNotFound(err) {
+		return resource.ExternalObservation{
+			ResourceExists: false,
+		}, nil
+	}
+
+	if err != nil {
+		return resource.ExternalObservation{}, errors.Wrapf(err, "Error getting IAMRole with name: %v", cr.Spec.RoleName)
+	}
+
+	cr.SetConditions(runtimev1alpha1.Available())
+
+	cr.Status.IAMRoleExternalState = cr.BuildExternalStateFromObservation(*observed.Role)
+
+	return resource.ExternalObservation{
+		ResourceExists: true,
+	}, nil
+}
+
+func (e *roleExternal) Create(ctx context.Context, mgd resource.Managed) (resource.ExternalCreation, error) {
+	cr, ok := mgd.(*v1alpha1.IAMRole)
+	if !ok {
+		return resource.ExternalCreation{}, errors.New(errObjectNotIAMRole)
+	}
+
+	cr.Status.SetConditions(runtimev1alpha1.Creating())
+
+	req := e.client.CreateRoleRequest(&awsiam.CreateRoleInput{
+		RoleName:                 aws.String(cr.Spec.RoleName),
+		AssumeRolePolicyDocument: aws.String(cr.Spec.AssumeRolePolicyDocument),
+		Description:              aws.String(cr.Spec.Description),
+	})
+	req.SetContext(ctx)
+
+	_, err := req.Send()
+
+	return resource.ExternalCreation{}, errors.Wrap(err, "failed to create the IAMRole resource")
+}
+
+func (e *roleExternal) Update(ctx context.Context, mgd resource.Managed) (resource.ExternalUpdate, error) {
+	cr, ok := mgd.(*v1alpha1.IAMRole)
+	if !ok {
+		return resource.ExternalUpdate{}, errors.New(errObjectNotIAMRole)
+	}
+
+	// TODO(soorena776): add more sophisticated Update logice, once we
+	// categorize immutable vs mutable fields (see #727)
+
+	req := e.client.UpdateRoleRequest(&awsiam.UpdateRoleInput{
+		RoleName:    aws.String(cr.Spec.RoleName),
+		Description: aws.String(cr.Spec.Description),
+	})
+	req.SetContext(ctx)
```

Comment:
It's not obvious to me whether this method omits logic to determine whether an update is required on purpose, or by accident. If we've decided to pay the cost of attempting an update each time I think we should include a comment explaining as much. See related issue https://github.com/crossplaneio/crossplane-runtime/issues/7.

Reply:
I'll remove this logic. This was the very first resource type that I implemented, and I was under the impression that there is a 1:1 mapping for each resource type's controller and client `Update` methods.